storage/storage_image.go

```diff
@@ -0,0 +1,554 @@
+func (s *storageImage) SupportsSignatures() error {
+	return nil
+}
+
+// PutBlob is used to both store filesystem layers and binary data that is part
+// of the image.  Filesystem layers are assumed to be imported in order, as
+// that is required by some of the underlying storage drivers.
+func (s *storageImage) PutBlob(stream io.Reader, blobinfo types.BlobInfo) (types.BlobInfo, error) {
+	actualSize := int64(-1)
+	digest := blobinfo.Digest
+	errorBlobInfo := types.BlobInfo{
+		Digest: "",
+		Size:   -1,
+	}
+	// Try to read an initial snippet of the blob.
+	header := make([]byte, 10240)
+	n, err := stream.Read(header)
+	if err != nil && err != io.EOF {
+		return errorBlobInfo, err
+	}
+	// Set up to read the whole blob (the initial snippet, plus the rest)
+	// while digesting it with sha256.
+	hasher := sha256.New()
+	hash := []byte{}
+	counter := ioutils.NewWriteCounter(hasher)
+	defragmented := io.MultiReader(bytes.NewBuffer(header[:n]), stream)
+	multi := io.TeeReader(defragmented, counter)
+	if (n > 0) && archive.IsArchive(header[:n]) {
+		// It's a filesystem layer.  If it's not the first one in the
+		// image, we assume that the most recently added layer is its
+		// parent.
+		parentLayer := ""
+		if len(s.BlobList) > 0 {
+			for _, blob := range s.BlobList {
+				if layerID, ok := s.Layers[blob]; ok {
+					parentLayer = layerID
+				}
+			}
+		}
+		// Now try to figure out if the identifier we have is an ID or
+		// something else, so that we can do collision detection
+		// correctly.
+		names := []string{}
+		id := blobinfo.Digest
+		if matches := idRegexp.FindStringSubmatch(blobinfo.Digest); len(matches) > 1 {
+			// Looks like a digest.  Use the value as a layer ID.
+			id = matches[len(matches)-1]
+		} else {
+			// If it isn't empty, it's just a name.
+			if blobinfo.Digest != "" {
+				names = append(names, blobinfo.Digest)
+				id = ""
+			}
+		}
+		if len(names) == 0 {
+			names = nil
+		}
+		// Attempt to create the identified layer and import its contents.
+		layer, err := s.store.PutLayer(id, parentLayer, names, "", true, multi)
+		if err != nil && err != storage.ErrDuplicateID {
+			logrus.Debugf("error importing layer blob %q: %v", blobinfo.Digest, err)
+			return errorBlobInfo, err
+		}
+		if err == storage.ErrDuplicateID {
+			// We specified an ID, and there's already a layer with
+			// the same ID.  Drain the input so that we can look at
+			// its length and digest.
+			_, err := io.Copy(ioutil.Discard, multi)
+			if err != nil && err != io.EOF {
+				logrus.Debugf("error digesting layer blob %q: %v", blobinfo.Digest, err)
+				return errorBlobInfo, err
+			}
+			hash = hasher.Sum(nil)
+		} else {
+			// Applied the layer, either with a specified ID or a
+			// new ID.  Note the size info and computed digest.
+			hash = hasher.Sum(nil)
+			layerdata := storageLayerMetadata{
+				ExpectedSize: blobinfo.Size,
+				Digest:       "sha256:" + hex.EncodeToString(hash[:]),
+				Size:         counter.Count,
```

Comment:
Probably a leftover from an earlier iteration.  I think it can be removed in favor of setting Size to -1 when we don't know, and have to recompute when reading its contents.